packages/menu/src/useMenu.ts

```diff
@@ -0,0 +1,659 @@
+    [defaultFocusedValue, isExpandedControlled, isFocusedValueControlled, onChange, values]
+  );
+
+  // How to control open/close with keydown
+  const handleMenuKeyDown = useCallback(
+    ({ key }) => {
+      if ([KEYS.ESCAPE, KEYS.TAB].includes(key)) {
+        const type = StateChangeTypes[key === KEYS.ESCAPE ? 'MenuKeyDownEscape' : 'MenuKeyDownTab'];
+
+        closeMenu(type);
+
+        const triggerEl = triggerRef?.current;
+
+        if (triggerEl && KEYS.TAB === key) {
+          triggerEl.focus();
+        }
+      }
+    },
+    [closeMenu, triggerRef]
+  );
+
+  const handleMenuBlur = useCallback(
+    event => {
+      const path = event.composedPath();
+
+      if (!path.includes(menuRef.current) && !path.includes(triggerRef.current)) {
+        closeMenu(StateChangeTypes.MenuBlur);
+      }
+    },
+    [closeMenu, menuRef, triggerRef]
+  );
+
+  const handleMenuMouseLeave = useCallback(() => {
+    dispatch({
+      type: StateChangeTypes.MenuMouseLeave,
+      payload: {}
+    });
+
+    onChange({ type: StateChangeTypes.MenuMouseLeave });
+  }, [onChange]);
+
+  /**
+   * Effects
+   */
+
+  /**
+   * Sets a state value to ensure 1 render happens after
+   * opening so a menu item can be focused.
+   */
+  useEffect(() => {
+    setMenuVisible(controlledIsExpanded);
+  }, [controlledIsExpanded, menuRef, environment]);
+
+  /**
+   * Respond to clicks outside the  open menu
+   */
+  useEffect(() => {
+    const doc = environment || document;
+
+    if (controlledIsExpanded) {
+      doc.addEventListener('click', handleMenuBlur, true);
+    } else if (!controlledIsExpanded) {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    }
+
+    return () => {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    };
+  }, [controlledIsExpanded, handleMenuBlur, environment]);
+
+  /**
+   * Return focus to trigger when menu closes
+   */
+  useEffect(() => {
+    if (controlledIsExpanded && !returnFocusRef.current) {
+      returnFocusRef.current = triggerRef.current;
+    }
+
+    if (!controlledIsExpanded && returnFocusRef.current) {
+      setTimeout(() => {
+        returnFocusRef.current!.focus();
+        returnFocusRef.current = null;
+      });
+    }
+  }, [triggerRef, controlledIsExpanded]);
+
+  /**
+   * Focus initial item when menu opens
+   */
+  useEffect(() => {
+    if (state.focusOnOpen && menuVisible && controlledFocusedValue && controlledIsExpanded) {
+      const ref = itemRefs[controlledFocusedValue]?.current;
+
+      ref && ref.focus();
+    }
+  }, [menuVisible, itemRefs, controlledFocusedValue, state.focusOnOpen, controlledIsExpanded]);
+
+  /**
+   * Prop getters
+   */
+
+  const getTriggerProps = useCallback<IUseMenuReturnValue<T, L>['getTriggerProps']>(
+    ({ onClick, onKeyDown, type = 'button', role = 'button', ...other } = {}) => ({
+      'data-garden-container-id': 'containers.menu.trigger',
+      'data-garden-container-version': PACKAGE_VERSION,
+      ref: triggerRef,
```

Comment:
I get LegacyRef vs RefObject issues here if `IUseMenuReturnValue` isn't generic. Even a nearly identical structure in Combobox doesn't have that issue. Open to ideas/pairing to get to the bottom of it.

Reply:
Looks like `useCombobox` doesn't see this error because of `getDownshiftTriggerProps` returns a props object with type `any`. So perhaps the type constructs I've set up here actually do make sense because otherwise the generic types wouldn't line up correctly.